File: Assets/BEHealthKit/Plugins/HealthData.cs
using UnityEngine;
using System;
using System.Collections.Generic;
using System.Xml;

namespace BeliefEngine.HealthKit
{

/*! @enum QuantityType
	@ingroup Enumerations
	@brief denotes whether a Quantity is cumulative, or discrete
 */
public enum QuantityType {
	cumulative,	discrete
}

/*! @brief Wrapper around HKQuantity.
 */
public class Quantity : System.Object {
	public string unit; /*!< the unit of this quantity, as a string */
	public double doubleValue; /*!< the value of this quantity, as a double */

	/*! @brief		The constructor used internally when reading health data.
		@param node	the XmlNode to create this object from.
	 */
	public Quantity(XmlNode node) {
		this.unit = node["unit"].InnerText;
		this.doubleValue = Convert.ToDouble(node["value"].InnerText);
	}

	/*! @brief				The default constructor.
		@param unitString	the string representation of the unit. For example, count, kg, or m/s^2.
							see: https://developer.apple.com/reference/healthkit/hkunit/1615733-unitfromstring
		@param value		the value of the quantity, as a double
	 */
	public Quantity(double value, string unitString) {
		this.unit = unitString;
		this.doubleValue = value;
	}

	/*! @brief convert to a reasonable string representation
	 */
	override public string ToString() {
		return string.Format("{0} {1}", this.doubleValue, this.unit);
	}
}

/*! @brief Wrapper around HKSample.
 */
public class Sample : System.Object {
	public DateTimeOffset startDate; /*!< the starting date of this sample */
	public DateTimeOffset endDate; /*!< the ending date of this sample */
	// sample type

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public Sample(XmlNode node) {
		this.startDate = DateTimeBridge.DateFromString(node["startDate"].InnerText);
		this.endDate = DateTimeBridge.DateFromString(node["endDate"].InnerText);
	}
}


/*! @brief Wrapper around HKQuantitySample.
 */
public class QuantitySample : Sample {
	public QuantityType quantityType; /*!< the aggregation style of this sample, either cumulative or discrete */
	public Quantity quantity; /*!< the quantity */

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public QuantitySample(XmlNode node) : base(node) {
		this.quantity = new Quantity(node["quantity"]);
		string aggregationStyle = node.SelectSingleNode("quantityType/aggregationStyle").InnerText;
		this.quantityType = (aggregationStyle == "cumulative") ? QuantityType.cumulative : QuantityType.discrete;
	}

	/*! @brief convert to a reasonable string representation
	 */
	override public string ToString() {
		return string.Format("[{0}-{1} : {2}]", this.startDate, this.endDate, this.quantity);
	}
}

/*! @brief Wrapper around HKCategorySample.
 */
public class CategorySample : Sample {
	/*! @brief 		the value of this sample
		@details	This is an int, and it's probably worth reading the [HKCategoryValueSleepAnalysis](https://developer.apple.com/library/ios/documentation/HealthKit/Reference/HealthKit_Constants/#//apple_ref/c/tdef/HKCategoryValueSleepAnalysis)
					documentation to understand what you're looking at. Basically a value of 0 means "in bed", and a 1 means "asleep".  These *will* overlap, assuming a good HealthKit citizen is writing the data.
	 */
	public int value;

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public CategorySample(XmlNode node) : base(node) {
		this.value = Int32.Parse(node["value"].InnerText);
	}

	/*! @brief convert to a reasonable string representation
	 */
	override public string ToString() {
		return string.Format("[{0}-{1} : {2}]", this.startDate, this.endDate, this.value);
	}
}

/*! @brief Wrapper around HKCorrelationSample.
 */
public class CorrelationSample : Sample {
	public string correlationType; /*!< TODO the correlation type. */
	public List<Sample> objects; /*!< the list of samples */

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public CorrelationSample(XmlNode node) : base(node) {
		this.correlationType = node["correlationType"].InnerText;
		this.objects = new List<Sample>();
		XmlNodeList sampleNodes = node.SelectNodes("objects");
		foreach (XmlNode sample in sampleNodes) {
			// can these be something other than Quantity Samples?
			this.objects.Add(new QuantitySample(sample));
		}
	}

	/*! @brief convert to a reasonable string representation
	 */
	override public string ToString() {
		string s = string.Format("[{0}:\n", correlationType);
		foreach (Sample sample in this.objects) {
			s = s + sample + "\n";
		}
		s = s + "]";
		return s;
	}
}

/*! @enum WorkoutEventType
	@ingroup Enumerations
	@brief denotes the type of Workout Event (Pause or Resume)
 */
public enum WorkoutEventType {
	Pause = 1,
	Resume
}

/*! @brief Wrapper around HKWorkoutEvent.
 */
public class WorkoutEvent : System.Object {
	public DateTimeOffset date; /*!< @brief time of the event */
	public WorkoutEventType eventType;	/*!< either Pause or Resume  */

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public WorkoutEvent(XmlNode node) {
		this.date = DateTimeBridge.DateFromString(node["date"].InnerText);
		this.eventType = (WorkoutEventType)Int32.Parse(node["eventType"].InnerText);
	}
}

/*! @brief Wrapper around HKWorkoutSample.
 */
public class WorkoutSample : Sample {
	public double duration;						/*!< @brief duration of the sample, in seconds */
	public Quantity totalDistance;				/*!< @brief total distance walked/run/etc. during the workout */
	public Quantity totalEnergyBurned;			/*!< @brief total energy burned during the workout */
	public WorkoutActivityType activityType;	/*!< @brief type of the workout */
	public List<WorkoutEvent> workoutEvents; 	/*!< @brief workout events contained in this sample */

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public WorkoutSample(XmlNode node) : base(node) {
		this.duration = Int32.Parse(node["duration"].InnerText);
		this.totalDistance = new Quantity(node["totalDistance"]);
		this.totalEnergyBurned = new Quantity(node["energyBurned"]);
		this.activityType = (WorkoutActivityType)Int32.Parse(node["activityType"].InnerText);

		this.workoutEvents = new List<WorkoutEvent>();
		XmlNodeList eventNodes = node.SelectNodes("events");
		foreach (XmlNode sample in eventNodes) {
			this.workoutEvents.Add(new WorkoutEvent(sample));
		}
	}

	// override public string ToString() {
	// 	return string.Format("[WORKOUT SAMPLE]");
	// }
}

/*! @brief Wrapper around HKCharacteristic.
 */
public class Characteristic : System.Object
{

}

/*! @brief Wrapper around HKBiologicalSexCharacteristic.
 */
public class BiologicalSexCharacteristic : Characteristic
{
	/*! @brief		the biological sex
		@details	This can either be NotSet, Male, Female, or Other.
	 */
	public BiologicalSex value;

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public BiologicalSexCharacteristic(XmlNode node) {
		this.value = (BiologicalSex)Int32.Parse(node.InnerText);
	}

	/*! @brief string representation. */
	override public string ToString() {
		return string.Format("[biological sex:{0}]", this.value);
	}
}

/*! @brief Wrapper around HKBloodTypeCharacteristic.
 */
public class BloodTypeCharacteristic : Characteristic
{
	/*! the blood type. NotSet is also a valid value. */
	public BloodType value;

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public BloodTypeCharacteristic(XmlNode node) {
		this.value = (BloodType)Int32.Parse(node.InnerText);
	}

	/*! @brief string representation. */
	override public string ToString() {
		return string.Format("[blood type:{0}]", this.value);
	}
}

/*! @brief Wrapper around HKDateOfBirthCharacteristic.
 */
public class DateOfBirthCharacteristic : Characteristic
{
	/*! the user's birthday. */
	public DateTimeOffset value;

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public DateOfBirthCharacteristic(XmlNode node) {
		this.value = DateTimeBridge.DateFromString(node.InnerText);
	}

	/*! @brief string representation. */
	override public string ToString() {
		return string.Format("[date of birth:{0}]", this.value);
	}
}

/*! @brief Wrapper around HKFitzpatrickSkinTypeObject.
 */
public class FitzpatrickSkinTypeCharacteristic : Characteristic
{
	/*! the skin type. NotSet is also a valid value. */
	public FitzpatrickSkinType value;

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public FitzpatrickSkinTypeCharacteristic(XmlNode node) {
		this.value = (FitzpatrickSkinType)Int32.Parse(node.InnerText);
	}

	/*! @brief string representation. */
	override public string ToString() {
		return string.Format("[skin type:{0}]", this.value);
	}
}

/*! @brief Wrapper around HKWheelchairUseObject.
 */
public class WheelchairUseCharacteristic : Characteristic
{
	/*! the skin type. NotSet is also a valid value. */
	public WheelchairUse value;

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public WheelchairUseCharacteristic(XmlNode node) {
		this.value = (WheelchairUse)Int32.Parse(node.InnerText);
	}

	/*! @brief string representation. */
	override public string ToString() {
		return string.Format("[wheelchair use:{0}]", this.value);
	}
}

/*! @brief Used to send HealthKit data back to Unity.
 */
public class HealthData : System.Object {
	
	public HKDataType datatype;  /*!< @brief the type of health data */
	public WorkoutActivityType workoutType;  /*!< @brief the workout type, if applicable */
	
	private XmlDocument xml;

	/*! @brief				The default constructor.
		@param xmlString	the XML string to create this object from.
	 */
	public HealthData(string xmlString) {
		xml = new XmlDocument();
		xml.LoadXml(xmlString);
		
		XmlNode node = xml.FirstChild["datatype"];
		if (node != null && node.InnerText != null) {
			this.datatype = (HKDataType)System.Enum.Parse(typeof(HKDataType), (string)node.InnerText);
		} else {
			Debug.LogError("datatype node is missing or invalid");
		}
		
		if (this.datatype == HKDataType.HKWorkoutTypeIdentifier) {
			node = xml.FirstChild["workoutType"];
			if (node != null && node.InnerText != null) this.workoutType = (WorkoutActivityType)Enum.ToObject(typeof(WorkoutActivityType), node.InnerText);
			else Debug.LogError("workoutType node is missing or invalid");
		}
	}

	/*! @brief		The name of the root node of the XML document.
		@details	This is used to determine what kind of HealthKit data to process as.
	 */
	public string RootName() {
		return xml.DocumentElement.Name;
	}

	/*! @brief parse XML containing QuantitySamples, and return a list. */
	public List<QuantitySample> ParseQuantitySamples() {
		XmlNodeList sampleNodes = xml.SelectNodes("/quantity/quantitySample");
		List<QuantitySample> samples = new List<QuantitySample>();
		foreach (XmlNode node in sampleNodes) {
			samples.Add(new QuantitySample(node));
		}

		return samples;
	}

	/*! @brief parse XML containing CategorySamples, and return a list. */
	public List<CategorySample> ParseCategorySamples() {
		XmlNodeList sampleNodes = xml.SelectNodes("/category/categorySample");
		List<CategorySample> samples = new List<CategorySample>();
		foreach (XmlNode node in sampleNodes) {
			samples.Add(new CategorySample(node));
		}

		return samples;
	}

	/*! @brief parse XML containing a combined total of QuantitySamples, and return a double. */
	public double ParseTotal() {
		XmlNode node = xml.SelectSingleNode("total/total");
		return Double.Parse(node.InnerText);
	}
	
	/*! @brief parse XML & determine if writing was a success. */
	public bool ParseSuccess() {
		XmlNode node = xml.SelectSingleNode("write/success");
		return bool.Parse(node.InnerText);
	}

	/*! @brief if there was an error, parse & return it; otherwise null. */
	public Error ParseError() {
		XmlNode node = xml.SelectSingleNode("write/error");
		if (node != null) return new Error(node);
		else return null;
	}

	/*! @brief parse XML containing CorrelationSamples, and return a list. */
	public List<CorrelationSample> ParseCorrelationSamples() {
		XmlNodeList sampleNodes = xml.SelectNodes("/correlation/correlationSample");
		List<CorrelationSample> samples = new List<CorrelationSample>();
		foreach (XmlNode node in sampleNodes) {
			samples.Add(new CorrelationSample(node));
		}

		return samples;
	}

	/*! @brief parse XML containing WorkoutSamples, and return a list. */
	public List<WorkoutSample> ParseWorkoutSamples() {
		XmlNodeList sampleNodes = xml.SelectNodes("/workout/workoutSample");
		List<WorkoutSample> samples = new List<WorkoutSample>();
		foreach (XmlNode node in sampleNodes) {
			samples.Add(new WorkoutSample(node));
		}

		return samples;
	}

	/*! @brief parse XML containing a Characteristic, and return it. */
	public Characteristic ParseCharacteristic() {
		foreach (XmlNode node in xml["characteristic"].ChildNodes) {
			switch (node.Name) {
			case "sex":
				return new BiologicalSexCharacteristic(node);
			case "bloodType":
				return new BloodTypeCharacteristic(node);
			case "skinType":
				return new FitzpatrickSkinTypeCharacteristic(node);
			case "wheelchairUse":
				return new WheelchairUseCharacteristic(node);
			case "DOB":
				return new DateOfBirthCharacteristic(node);
			case "datatype":
				continue;
			default:
				Debug.LogErrorFormat("Error: unrecognized characteristic '{0}'", node.Name);
				break;
		}
		}
		
		return null;
	}

	/*! @brief parse XML containing pedometer data, and return it. */
	public List<PedometerData> ParsePedometerData() {
		XmlNodeList sampleNodes = xml.SelectNodes("/pedometer/pedometerData");
		List<PedometerData> samples = new List<PedometerData>();
		foreach (XmlNode node in sampleNodes) {
			samples.Add(new PedometerData(node));
		}

		return samples;
	}
}

/*! @brief Wrapper around CMPedometerData.
 */
public class PedometerData : System.Object {
	public DateTimeOffset startDate; /*!< the starting date of this sample */
	public DateTimeOffset endDate; /*!< the ending date of this sample */
	public int numberOfSteps; /*!< the number of steps taken in this sample */
	// sample type

	/*! @brief		The default constructor.
		@param node	the XmlNode to create this object from.
	 */
	public PedometerData(XmlNode node) {
		this.startDate = DateTimeBridge.DateFromString(node["startDate"].InnerText);
		this.endDate = DateTimeBridge.DateFromString(node["endDate"].InnerText);
		this.numberOfSteps = Convert.ToInt32(node["numberOfSteps"].InnerText);
	}
}

}

File: Assets/BEHealthKit/iOS/HealthData.m
//
//  HealthData.m
//  Unity-iPhone
//
//  Created by greay on 3/28/15.
//
//

#import "HealthData.h"
#import <HealthKit/HealthKit.h>
#import "BEHealthKit.h"
#import "XMLDictionary/XMLDictionary.h"
#import "NSDate+bridge.h"

// ---------------------
// MARK: serialization
// ---------------------

@implementation HKObjectType (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[@"identifier"] = self.identifier;
	return dict;
}

@end


@implementation HKSampleType (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	// nothing?
	return dict;
}

@end


@implementation HKQuantityType (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	dict[@"aggregationStyle"] = (self.aggregationStyle == HKQuantityAggregationStyleCumulative) ? @"cumulative" : @"discrete";
	return dict;
}

@end


@implementation HKCorrelationType (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	// nothing?
	return dict;
}

@end

@implementation HKQuantity (serialization)

- (id)be_serializableWithUnit:(HKUnit *)unit {
//	NSArray *units = @[[HKUnit countUnit], [HKUnit percentUnit], [HKUnit gramUnit], [HKUnit kilocalorieUnit],
//					   [[HKUnit countUnit] unitDividedByUnit:[HKUnit minuteUnit]],
//					   [HKUnit degreeCelsiusUnit], [HKUnit millimeterOfMercuryUnit],
//					   [[HKUnit gramUnit] unitDividedByUnit:[HKUnit literUnit]],
//					   [HKUnit siemenUnit],
//					   [[HKUnit literUnit] unitDividedByUnit:[HKUnit hourUnit]]
//					   ];
	if ([self isCompatibleWithUnit:unit]) {
		NSMutableDictionary *dict = [NSMutableDictionary dictionary];
		dict[@"unit"] = [unit unitString];
		dict[@"value"] = @([self doubleValueForUnit:unit]);
		
		return dict;
	}
	NSLog(@"error; don't know which unit to use!");
	return nil;
}

@end


@implementation HKObject (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	if (self.metadata) {
		NSMutableDictionary *metadata = [@{} mutableCopy];
		for (NSString *key in self.metadata) {
			if ([self.metadata[key] isKindOfClass:[NSString class]] || [self.metadata[key] isKindOfClass:[NSNumber class]]) {
				metadata[key] = self.metadata[key];
			}
			else if ([self.metadata[key] isKindOfClass:[NSDate class]]) {
				metadata[key] = [(NSDate *)self.metadata[key] bridgeToken];
			}
			else {
				NSLog(@"error:don't know how to handle metadata[%@] = %@ (%@)", key, self.metadata[key], [self.metadata[key] class]);
			}
		}
		dict[@"metadata"] = metadata;
		
	}
	return dict;
}

@end



@implementation HKSample (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	dict[@"startDate"] = [self.startDate bridgeToken];
	dict[@"endDate"] = [self.endDate bridgeToken];
	dict[@"sampleType"] = [self.sampleType be_serializable];
	return dict;
}

@end


@implementation HKQuantitySample (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	dict[@"quantityType"] = [self.quantityType be_serializable];
	
	HKUnit *unit = [[BEHealthKit sharedHealthKit] defaultUnitForSampleType:self.quantityType];
	dict[@"quantity"] = [self.quantity be_serializableWithUnit:unit];
	
	return dict;
}

@end


@implementation HKCorrelation (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
#warning this doesnt seem to do much of anything!
	dict[@"correlationType"] = [self.correlationType be_serializable];
	
	NSMutableArray *objects = [NSMutableArray array];
	for (HKSample *object in self.objects) {
		#warning what class are these? 
		[objects addObject:[object be_serializable]];
	}
	dict[@"objects"] = objects;
	
	return dict;
}

@end


@implementation HKWorkoutEvent (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[@"date"] = [self.date bridgeToken];
	dict[@"eventType"] = @(self.type);
	
	return dict;
}

@end


@implementation HKWorkout (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	dict[@"duration"] = @(self.duration);
	dict[@"totalDistance"] = [self.totalDistance be_serializableWithUnit:[HKUnit mileUnit]];
	dict[@"energyBurned"] = [self.totalEnergyBurned be_serializableWithUnit:[HKUnit kilocalorieUnit]];
	dict[@"activityType"] = @(self.workoutActivityType);
	
	NSMutableArray *events = [NSMutableArray array];
	for (HKWorkoutEvent *event in self.workoutEvents) {
		[events addObject:[event be_serializable]];
	}
	dict[@"events"] = events;
	
	return dict;
}

@end


@implementation HKCategoryType (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	return dict;
}

@end


@implementation HKCategorySample (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [super be_serializable];
	dict[@"categoryType"] = [self.categoryType be_serializable];
	dict[@"value"] = @(self.value);
	return dict;
}

@end


@implementation CMPedometerData (serialization)

- (id)be_serializable {
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[@"startDate"] = [self.startDate bridgeToken];
	dict[@"endDate"] = [self.endDate bridgeToken];
	dict[@"numberOfSteps"] = self.numberOfSteps;
	return dict;
}

@end

// ---------------------
// MARK: XML
// ---------------------


@implementation HealthData

+ (NSString *)XMLFromQuantitySamples:(NSArray *)quantitySamples datatype:(NSString *)datatype
{
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[XMLDictionaryNodeNameKey] = @"quantity";
	dict[@"datatype"] = datatype;
	
	NSMutableArray *samples = [NSMutableArray array];
	for (HKQuantitySample *sample in quantitySamples) {
		[samples addObject:[sample be_serializable]];
	}
	dict[@"quantitySample"] = samples;
	
	return [dict XMLString];
//	return [self plistXML:dict];
}

+ (NSString *)XMLFromCombinedTotal:(double)total
{
	NSDictionary *dict = @{@"total":@(total), XMLDictionaryNodeNameKey:@"total"};
	return [dict XMLString];
}

+ (NSString *)XMLFromCategorySamples:(NSArray *)categorySamples datatype:(NSString *)datatype
{
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[XMLDictionaryNodeNameKey] = @"category";
	dict[@"datatype"] = datatype;
	
	NSMutableArray *samples = [NSMutableArray array];
	for (HKCategorySample *sample in categorySamples) {
		[samples addObject:[sample be_serializable]];
	}
	dict[@"categorySample"] = samples;
	
	return [dict XMLString];
}

+ (NSString *)XMLFromCorrelationSamples:(NSArray *)correlationSamples datatype:(NSString *)datatype
{
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[XMLDictionaryNodeNameKey] = @"correlation";
	dict[@"datatype"] = datatype;
	
	NSMutableArray *samples = [NSMutableArray array];
	for (HKCorrelation *sample in correlationSamples) {
		[samples addObject:[sample be_serializable]];
	}
	dict[@"correlationSample"] = samples;
	
	return [dict XMLString];
}

+ (NSString *)XMLFromWorkoutSamples:(NSArray *)workoutSamples workoutType:(int)workoutType
{
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[XMLDictionaryNodeNameKey] = @"workout";
	dict[@"datatype"] = HKWorkoutTypeIdentifier;
	dict[@"workoutType"] = @(workoutType);
	
	NSMutableArray *samples = [NSMutableArray array];
	for (HKWorkout *sample in workoutSamples) {
		[samples addObject:[sample be_serializable]];
	}
	dict[@"workoutSample"] = samples;
	
	return [dict XMLString];
}


+ (NSString *)XMLFromCharacteristic:(id)characteristic datatype:(NSString *)datatype
{
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[XMLDictionaryNodeNameKey] = @"characteristic";
	dict[@"datatype"] = datatype;
	
	if ([characteristic isKindOfClass:[HKBiologicalSexObject class]]) {
		dict[@"sex"] = @([(HKBiologicalSexObject *)characteristic biologicalSex]);
	} else if ([characteristic isKindOfClass:[HKBloodTypeObject class]]) {
		dict[@"bloodType"] = @([(HKBloodTypeObject *)characteristic bloodType]);
	} else if ([characteristic isKindOfClass:[HKFitzpatrickSkinTypeObject class]]) {
		dict[@"skinType"] = @([(HKFitzpatrickSkinTypeObject *)characteristic skinType]);
	} else if ([characteristic isKindOfClass:[HKWheelchairUseObject class]]) {
		dict[@"wheelchairUse"] = @([(HKWheelchairUseObject *)characteristic wheelchairUse]);
	} else if ([characteristic isKindOfClass:[NSDate class]]) {
		dict[@"DOB"] = [(NSDate *)characteristic bridgeToken];
	} else {
		NSLog(@"Error; unrecognized characteristic:%@", [characteristic class]);
		return nil;
	}
	
	return [dict XMLString];
}

+ (NSString *)XMLFromPedometerData:(CMPedometerData *)data
{
	NSMutableDictionary *dict = [NSMutableDictionary dictionary];
	dict[@"pedometerData"] = [data be_serializable];
	dict[XMLDictionaryNodeNameKey] = @"pedometer";

	return [dict XMLString];
}


+ (NSString *)plistXML:(NSDictionary *)plist
{
	NSError *error;
	NSData *data = [NSPropertyListSerialization dataWithPropertyList:plist format:NSPropertyListXMLFormat_v1_0 options:0 error: &error];
	if (data == nil) {
		NSLog (@"error serializing to xml: %@", error);
		return nil;
	}
	return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

@end

